### PR TITLE
[ES barcamp] JAMES-2202 Brings alias into Apache James

### PR DESCRIPTION
**Pull request by chibenwa**
We create a read index and a write index.

(This unlock potential scenario like update old and new index and read only from old index, which makes handling concurrency on re-index easier)

Two refactoring are provided:
 - Separation of concerns between configuration parsing and Guice injections
 - Improve IndexCreationFactory by making it a real factory (avoids static methods/arguments multiplications)

Next step would be to use alias for document re-indexing.

### CODE REVIEW

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java [hunk 1]
```diff
@@ -39,6 +42,44 @@
     public ExpectedException expectedException = ExpectedException.none();
 
     @Test
+    public void provideIndexNameShouldRetrievedConfiguredIndexName() throws ConfigurationException {
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java]: s/Retreived/Retrieve

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java [hunk 2]
```diff
@@ -80,6 +81,44 @@ public void provideIndexNameShouldReturnDefaultIndexNameWhenError() throws Confi
     }
 
     @Test
+    public void provideAliasNameShouldRetrievedConfiguredAliasName() throws ConfigurationException {
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java]: s/Retrieved/Retrieve

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+
+    @Test
+    public void getNbReplicaShouldReturnDefaultValueWhenMissing() throws ConfigurationException {
+        PropertiesConfiguration configuration = new PropertiesConfiguration();
+        configuration.addProperty("elasticsearch.hosts", "127.0.0.1");
+
+        ElasticSearchConfiguration elasticSearchConfiguration = ElasticSearchConfiguration.fromProperties(configuration);
+
+        assertThat(elasticSearchConfiguration.getNbShards())
+            .isEqualTo(ElasticSearchConfiguration.DEFAULT_NB_SHARDS);
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java]: DEFAULT_NB_REPLICA

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        configuration.addProperty("elasticsearch.hosts", "127.0.0.1");
+
+        ElasticSearchConfiguration elasticSearchConfiguration = ElasticSearchConfiguration.fromProperties(configuration);
+
+        assertThat(elasticSearchConfiguration.getWriteAliasName())
+            .isEqualTo(MailboxElasticsearchConstants.DEFAULT_MAILBOX_WRITE_ALIAS);
+    }
+
+    @Test
+    public void getReadAliasNameShouldReturnConfiguredValueWhenTrue() throws ConfigurationException {
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java]: getIndexAttachment

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+        configuration.addProperty("elasticsearch.hosts", "127.0.0.1");
+
+        ElasticSearchConfiguration elasticSearchConfiguration = ElasticSearchConfiguration.fromProperties(configuration);
+
+        assertThat(elasticSearchConfiguration.getIndexAttachment())
+            .isEqualTo(IndexAttachments.YES);
+    }
+
+    @Test
+    public void getReadAliasNameShouldReturnConfiguredValueWhenFalse() throws ConfigurationException {
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java]: getIndexAttachment

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+        configuration.addProperty("elasticsearch.port", port);
+
+        ElasticSearchConfiguration elasticSearchConfiguration = ElasticSearchConfiguration.fromProperties(configuration);
+
+        assertThat(elasticSearchConfiguration.getHosts())
+            .containsOnly(Host.from(hostname, port));
+    }
+
+    @Test
+    public void getIndexAttachmentShouldReturnDefaultValueWhenMissing() throws ConfigurationException {
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java]: already tested

Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java]: No, it was tested with true or false but not without value.

I located it with other methods for easier reading ;-)

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java (hunk 2 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java]: Removed

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java (hunk 1 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModuleTest.java]: Removed

**Comment by chibenwa** [comment]
```
[c08a48cbffcc94a7e07af052d1cf7c74151268cf] [INFO] Results:
[c08a48cbffcc94a7e07af052d1cf7c74151268cf] [INFO] 
[c08a48cbffcc94a7e07af052d1cf7c74151268cf] [ERROR] Errors: 
[c08a48cbffcc94a7e07af052d1cf7c74151268cf] [ERROR]   CassandraMessageMapperTest.org.apache.james.mailbox.cassandra.mail.CassandraMessageMapperTest » ContainerLaunch
[c08a48cbffcc94a7e07af052d1cf7c74151268cf] [INFO] 
[c08a48cbffcc94a7e07af052d1cf7c74151268cf] [ERROR] Tests run: 568, Failures: 0, Errors: 1, Skipped: 1
```

Not related failure...

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/AliasName.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        return value;
+    }
+
+    @Override
+    public final boolean equals(Object o) {
+        if (o instanceof AliasName) {
+            AliasName aliasName = (AliasName) o;
+
+            return Objects.equals(this.value, aliasName.value)
+                ;
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/AliasName.java]: why this newline before semi-column ?

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/NodeMappingFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import org.junit.rules.TemporaryFolder;
+
+public class NodeMappingFactoryTest {
+    public static final String MESSAGE = "message";
+    public static final IndexName INDEX_NAME = new IndexName("index");
+    public static final AliasName ALIAS_NAME = new AliasName("alias");
+    public static final TypeName TYPE_NAME = new TypeName("type");
+
+    private TemporaryFolder temporaryFolder = new TemporaryFolder();
+    private EmbeddedElasticSearch embeddedElasticSearch= new EmbeddedElasticSearch(temporaryFolder, INDEX_NAME);
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/NodeMappingFactoryTest.java]: missing space before `=`

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/MailboxElasticsearchConstants.java [hunk 9]
```diff
@@ -19,10 +19,12 @@
 
 package org.apache.james.mailbox.elasticsearch;
 
+import org.apache.james.backends.es.AliasName;
 import org.apache.james.backends.es.IndexName;
 import org.apache.james.backends.es.TypeName;
 
 public interface MailboxElasticsearchConstants {
+    AliasName DEFAULT_MAILBOX_ALIAS = new AliasName("mailboxAlias");
```
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/MailboxElasticsearchConstants.java]: I would name that one mailbox and I would call the real index depending on its actual content (v1, v2, or anything that would help identifying its content)

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/AliasInjectionConstants.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.backends.es;
+
+public interface AliasInjectionConstants {
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/AliasInjectionConstants.java]: Constants or ElasticSearchConstants would be better IMO

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java [hunk 11]
```diff
@@ -31,33 +32,65 @@
...
 
-    public static Client createIndexAndAlias(Client client, IndexName indexName, AliasName aliasName, int nbShards, int nbReplica) {
-        try {
-            return createIndexAndAlias(client, indexName, aliasName, generateSetting(nbShards, nbReplica));
-        } catch (IOException e) {
-            LOGGER.error("Error while creating index : ", e);
-            return client;
-        }
+    private IndexName indexName;
+    private ImmutableList.Builder<AliasName> aliases;
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/IndexCreationFactory.java]: ArrayList is more than enough in that context (you never publish the list)

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java [hunk 12]
```diff
@@ -58,8 +58,10 @@
     public void setup() throws IOException {
         node = embeddedElasticSearch.getNode();
         TestingClientProvider clientProvider = new TestingClientProvider(node);
-        IndexCreationFactory.createIndexAndAlias(clientProvider.get(),
-            INDEX_NAME, ALIAS_NAME);
+        new IndexCreationFactory()
+            .onIndex(INDEX_NAME)
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java]: I don't understand `onIndex`, what means on ?

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTest.java [hunk 13]
```diff
@@ -109,10 +109,11 @@ protected void await() {
     @Override
     protected void initializeMailboxManager() throws Exception {
         Client client = NodeMappingFactory.applyMapping(
-            IndexCreationFactory.createIndexAndAlias(
-                new TestingClientProvider(embeddedElasticSearch.getNode()).get(),
-                MailboxElasticsearchConstants.DEFAULT_MAILBOX_INDEX,
-                MailboxElasticsearchConstants.DEFAULT_MAILBOX_ALIAS),
+            new IndexCreationFactory()
+                .onIndex(MailboxElasticsearchConstants.DEFAULT_MAILBOX_INDEX)
+                .addAlias( MailboxElasticsearchConstants.DEFAULT_MAILBOX_READ_ALIAS)
```
Comment (mbaechler) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTest.java]: you should not put a space before MailboxElasticsearchConstants

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchConfiguration.java [hunk 14]
```diff
@@ -19,13 +19,192 @@
...
+        Optional<String> monoHostAddress = Optional.ofNullable(
+            propertiesReader.getString(ELASTICSEARCH_MASTER_HOST, null));
+        Optional<Integer> monoHostPort = Optional.ofNullable(
+            propertiesReader.getInteger(ELASTICSEARCH_PORT, null));
+        Optional<String> multiHosts = Optional.ofNullable(
+            propertiesReader.getString(ELASTICSEARCH_HOSTS, null));
+
+        validateHostsConfigurationOptions(monoHostAddress, monoHostPort, multiHosts);
+
+        if (monoHostAddress.isPresent()) {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchConfiguration.java]: masterHost ?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/AliasName.java (hunk 7 above)
Comment (chibenwa) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/AliasName.java]: Template approximation

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java (hunk 12 above)
Comment (chibenwa) [reply backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java]: Ít more useIndex I guess...

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/MailboxElasticsearchConstants.java (hunk 9 above)
Comment (chibenwa) [reply mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/MailboxElasticsearchConstants.java]: When you have a read and a write alias it makes sense IMO to have them called **mailboxWriteAlias** and **mailboxReadAlias**.

Regarding the index, I will add a version on it, if it please you.

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java [hunk 15]
```diff
@@ -32,6 +32,7 @@
 import org.apache.james.mailbox.elasticsearch.MailboxElasticSearchConstants;
 import org.apache.james.util.Host;
 import org.junit.Test;
+import org.testcontainers.shaded.com.google.common.collect.ImmutableList;
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/mailbox/ElasticSearchConfigurationTest.java]: bad import

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/IndexCreationFactoryTest.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+package org.apache.james.backends.es;
+
+import org.apache.james.backends.es.utils.TestingClientProvider;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TemporaryFolder;
+
+public class IndexCreationFactoryTest {
```
Comment (aduprat) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/IndexCreationFactoryTest.java]: You are not testing `Preconditions`

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java [hunk 17]
```diff
@@ -56,105 +57,83 @@
...
-            .get();
+    private ElasticSearchConfiguration getElasticSearchConfiguration(PropertiesProvider propertiesProvider) throws ConfigurationException {
+        try {
+            PropertiesConfiguration configuration = propertiesProvider.getConfiguration(ELASTICSEARCH_CONFIGURATION_NAME);
+            return ElasticSearchConfiguration.fromProperties(configuration);
+        } catch (FileNotFoundException e) {
+            LOGGER.warn("Could not find " + ELASTICSEARCH_CONFIGURATION_NAME + " configuration file. Using 127.0.0.1:9300 as contact point");
+            PropertiesConfiguration configuration = new PropertiesConfiguration();
+            configuration.addProperty(ElasticSearchConfiguration.ELASTICSEARCH_HOSTS, LOCALHOST);
+            return ElasticSearchConfiguration.fromProperties(configuration);
```
Comment (aduprat) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/ElasticSearchMailboxModule.java]: We may have a `ElasticSearchConfiguration.DEFAULT` constant.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**